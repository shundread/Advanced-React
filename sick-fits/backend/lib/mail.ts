import { createTransport } from "nodemailer";

const transport = createTransport({
  host: process.env.MAIL_HOST,
  port: process.env.MAIL_PORT,
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function passwordResetEmailBody(token: string) {
  return `
    <div style="
      border: 1px solid black;
      padding: 20px;
      font-family: sans-serif;
      line-height: 2;
      font-size: 20px;
    ">
      <p>Your password reset link is here, sir:</p>
      <p>
        <a href="${process.env.FRONTEND_URL}/reset?token=${token}">Go to page</a>
      </p>
      <p>Greetings, Foobar</p>
    </div>
  `;
}

export async function sendPasswordResetEmail({
  resetToken,
  to,
}: {
  resetToken: string;
  to: string;
}) {
  transport.sendMail({
    to,
    from: process.env.MAIL_USER,
    subject: "Your password reset token, sir!",
    html: passwordResetEmailBody(resetToken),
  });
}
