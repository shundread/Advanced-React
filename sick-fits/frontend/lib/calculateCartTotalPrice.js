export function calculateCartTotalPrice(cart) {
  return cart.reduce((tally, cartItem) => {
    // The product could have been deleted while still being in the cart
    if (!cartItem.product) return tally;
    return tally + cartItem.product.price * cartItem.quantity;
  }, 0);
}
