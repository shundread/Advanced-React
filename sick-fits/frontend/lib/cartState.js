import { createContext, useCallback, useContext, useState } from "react";

const LocalStateContext = createContext();

export function CartStateProvider({ children }) {
  // This is our own custom provider we will store data (state) and functionality (updaters)
  // in here and anyone can access it via the consumer
  const [cartOpen, setCartOpen] = useState(false);
  const toggleCart = useCallback(() => setCartOpen((open) => !open), []);
  const openCart = useCallback(() => setCartOpen(true), []);
  const closeCart = useCallback(() => setCartOpen(false), []);

  return (
    <LocalStateContext.Provider
      value={{ cartOpen, setCartOpen, toggleCart, openCart, closeCart }}
    >
      {children}
    </LocalStateContext.Provider>
  );
}

// Make a custom hook for accessing the cart local state
export function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}
