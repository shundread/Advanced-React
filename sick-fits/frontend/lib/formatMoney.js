const formatter = Intl.NumberFormat("en-US", {
  style: "currency",
  currency: "USD",
  minimumFractionDigits: 2,
});

export function formatMoney(amount) {
  return formatter.format(amount / 100);
}
