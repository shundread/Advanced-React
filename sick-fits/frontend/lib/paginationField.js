import { PAGINATION_QUERY } from "../components/Pagination";

export function paginationField() {
  return {
    keyArgs: false, // Tells Apollo that we will take care of everything
    read(existing = [], { args, cache }) {
      console.log({ existing, args, cache });
      const { skip, first } = args;

      // Read the number of items on the page from the cache
      const paginationData = cache.readQuery({ query: PAGINATION_QUERY });

      const itemCount = paginationData?._allProductsMeta?.count;
      const page = Math.ceil(skip / first) + 1;
      const pages = Math.ceil(itemCount / first);

      // Asks the read function for those items
      const items = existing.slice(skip, skip + first).filter((x) => x);
      console.log(
        "First is",
        first,
        "\nExisting is\n",
        existing,
        "\nItems is\n",
        items
      );

      // WTF
      if (items.length && items.length !== first && page === pages) {
        return items;
      }

      if (items.length !== first) {
        // We don't have the right amount of items, we must go to the network to fetch tem
        return false;
      }

      // If there are items, just return them from the cache and we don't need to go to the network
      if (items.length) {
        console.log(`There are ${items.length} items in the cache.`);
        return items;
      }

      return false;

      // We can either do one of two things:
      // 1) Return the items because they are already in the cache
      // or
      // 2) Return false from here (network request)
    },
    merge(existing, incoming, { args }) {
      // This runs when the Apollo client comes back from the network with our items
      const { skip } = args;
      const merged = existing ? existing.concat() : [];
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      console.log("Merged is\n", merged);
      return merged;
    },
  };
}
