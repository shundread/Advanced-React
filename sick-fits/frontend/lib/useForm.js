import { useEffect, useState } from "react";

export function useForm(initialState = {}) {
  // Create a state object for our inputs
  const [inputs, setInputs] = useState(initialState);

  useEffect(() => setInputs(initialState), [initialState]);

  function handleChange(event) {
    setInputs({
      ...inputs,
      [event.target.name]: parseEventValue(event),
    });
  }

  function clearForm() {
    const blankState = Object.fromEntries(
      Object.keys(inputs).map((key) => [key, ""])
    );
    setInputs(blankState);
  }

  function resetForm() {
    setInputs(initialState);
  }

  return {
    inputs,
    clearForm,
    handleChange,
    resetForm,
  };
}

function parseEventValue(event) {
  if (event.target.type === "number") return parseInt(event.target.value);
  if (event.target.type === "file") return event.target.files[0]?.name || ""; // event.target.files;
  return event.target.value;
}
