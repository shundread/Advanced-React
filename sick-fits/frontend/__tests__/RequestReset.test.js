import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import userEvent from "@testing-library/user-event";
import { fakeUser } from "../lib/testUtils";
import {
  RequestReset,
  REQUEST_RESET_PASSWORD_MUTATION,
} from "../components/RequestReset";

const MockedUser = fakeUser();

const mocks = [
  {
    request: {
      query: REQUEST_RESET_PASSWORD_MUTATION,
      variables: { email: MockedUser.email },
    },
    result: {
      data: { sendUserPasswordResetLink: null },
    },
  },
];

describe("<RequestReset />", () => {
  it("Renders and matches the snapshot", () => {
    const { container } = render(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("Calls the mutation when submitted", async () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );

    // Type into the email box
    userEvent.type(screen.getByPlaceholderText(/email/i), MockedUser.email);
    userEvent.click(screen.getByText("Request Password Reset!"));

    const success = await screen.findByText(/success/i);
    expect(success).toBeInTheDocument();
  });
});
