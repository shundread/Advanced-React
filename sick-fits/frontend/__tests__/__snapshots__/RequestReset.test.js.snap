// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset /> Renders and matches the snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 enVODd"
    method="POST"
  >
    <h2>
      Forgot Your Password?
    </h2>
    <fieldset>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Your Email Address"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Request Password Reset!
      </button>
    </fieldset>
  </form>
</div>
`;
